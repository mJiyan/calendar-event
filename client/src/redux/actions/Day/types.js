const SET_DAY_SELECTED = 'DAY/SET_DAY_SELECTED';

const actionTypes = {
  SET_DAY_SELECTED,
};

export default actionTypes;
