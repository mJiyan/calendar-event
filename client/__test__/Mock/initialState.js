import dayjs from 'dayjs';

const initialState = {
  EventReducer: {
    events: [
      {
        _id: '619f90f7abfdf7635dc91f43',
        day: 1636531800000,
        description: 'Meet with friends',
        id: '619f90f7abfdf7635dc91f43',
        label: 'purple',
        title: 'Meeting',
      },
      {
        _id: '61cd677c3254c8001e8f0769',
        day: dayjs(),
        description: 'Buy new notebook',
        id: '61cd677c3254c8001e8f0769',
        label: 'blue',
        title: 'Notebook',
      },
    ],
    event: {},
    selectedEvent: {
      _id: "61cd8200bc6749001f5bf489",
      title: "Meeting",
      description: "Meet with friends",
      label: "purple",
      day: 1638392400000,
      __v: 0,
      id: "61cd8200bc6749001f5bf489"
    },
  },
  MonthReducer: {
    month: 1
  },
  DayReducer: {
    day: dayjs()
  },
  ModalReducer: {
    modal: false
  }
};


export default initialState;