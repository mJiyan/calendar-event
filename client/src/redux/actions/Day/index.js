import actionTypes from './types';

export const setDaySelected = (day) => async (dispatch) => {
  dispatch({
    type: actionTypes.SET_DAY_SELECTED,
    payload: day,
  });
};
