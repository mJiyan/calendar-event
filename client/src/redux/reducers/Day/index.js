import actionTypes from '../../actions/Day/types';

const initialState = {
  day: {},
};

const DayReducer = (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.SET_DAY_SELECTED:
      return { ...state, day: action.payload };
    default:
      return state;
  }
};

export default DayReducer;
