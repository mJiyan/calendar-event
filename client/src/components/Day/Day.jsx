import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import dayjs from 'dayjs';
import * as dayActions from '../../redux/actions/Day';
import * as modalActions from '../../redux/actions/Modal';
import * as eventActions from '../../redux/actions/Event';

const Day = ({ day, rowIndex }) => {
  const [dayEvents, setDayEvents] = useState([]);

  const dispatch = useDispatch();
  const savedEvents = useSelector((state) => state.EventReducer.events);

  useEffect(() => {
    const events = savedEvents.filter(
      (evt) => dayjs(evt.day).format('DD-MM-YY') === day.format('DD-MM-YY'),
    );
    setDayEvents(events);
  }, [savedEvents, day]);

  const isItPresentDay = () =>
    day.format('DD-MM-YY') === dayjs().format('DD-MM-YY')
      ? 'bg-blue-600 text-white rounded-full w-7'
      : '';

  dayEvents.sort((a, b) => {
    const keyA = new Date(a.day);
    const keyB = new Date(b.day);
    // Compare the 2 dates
    if (keyA < keyB) return -1;
    if (keyA > keyB) return 1;
    return 0;
  });

  return (
    <div className="border border-gray-200 flex flex-col">
      <header className="flex flex-col items-center">
        {rowIndex === 0 && <p className="text-sm p-1 mt-1">{day.format('ddd').toUpperCase()}</p>}
        <p className={`text-sm p-1 my-1 text-center ${isItPresentDay()}`}>{day.format('DD')}</p>
      </header>
      <div
        className="flex-1 cursor-pointer"
        onClick={() => {
          dispatch(dayActions.setDaySelected(day));
          dispatch(modalActions.setShowEventModal(true));
        }}
      >
        {dayEvents.map((evt, idx) => (
          <div
            key={idx}
            onClick={() => dispatch(eventActions.setSelectedEvent(evt))}
            className={`bg-${evt.label}-200 p-1 mr-3 text-gray-600 text-sm rounded mb-1 truncate`}
          >
            {evt.title}
          </div>
        ))}
      </div>
    </div>
  );
};

export default Day;
