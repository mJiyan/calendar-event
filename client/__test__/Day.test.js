import React from 'react';
import Day from '../src/components/Day/Day';
import renderer from 'react-test-renderer';
import { mount, render, shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import toJson from 'enzyme-to-json'; //added this line
import dayjs from 'dayjs';
import initialState from './Mock/initialState';


const mockStore = configureMockStore([thunk]);

const day = dayjs(1636531800000); // date -> 2021-11-10
const presentDay = dayjs(); // today
const rowDay = dayjs(1638392400000) // date -> 2021-12-02

const store = mockStore(initialState);

const wrapper = renderer.create(
  <Provider store={store}>
    <Day day={day} key={1} rowIndex={1} />
  </Provider>,
);


const wrapperMount = mount(
  <Provider store={store}>
    <Day day={day} key={1} rowIndex={1} />
  </Provider>,
);

const rowWrapperMount = mount(
  <Provider store={store}>
    <Day day={rowDay} key={1} rowIndex={0} />
  </Provider>,
);


const presentDayMount = mount(
  <Provider store={store}>
    <Day day={presentDay} key={0} rowIndex={0} />
  </Provider>,
);


describe('rendering components', () => {

  test('it should match with the month component', () => {
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('it should show the 10th day of month according to the date value', () => {
    const currentDayParagraph = wrapperMount.find('[data-testid="day-value-10-11-21"]');
    expect(toJson(currentDayParagraph).children[0]).toEqual('10');
  });

  test('it should show the row text value', () => {
    const rowText = rowWrapperMount.find(`[day-testid="row-value-02-12-21"]`);
    expect(rowText.length).toEqual(1);
  });


  test('it should show the present day in special style', () => {
    const currentDayParagraph = presentDayMount.find(`[data-testid="day-value-${presentDay.format('DD-MM-YY')}"]`);
    expect(toJson(currentDayParagraph).props.className).toEqual("text-sm p-1 my-1 text-center bg-blue-600 text-white rounded-full w-7");
  });


  test('it should execute dispatch properly for event reducer and day reducer', () => {

    const selectDay = wrapperMount.find(`[data-testid="select-day-${day.format('DD-MM-YY')}"]`);
    selectDay.simulate('click');


    const actions = store.getActions();
    const expectedPayload = [
      {
        type: 'DAY/SET_DAY_SELECTED',
        payload: dayjs("2021-11-10T08:10:00.000Z")
      },
      { type: 'MODAL/SET_MODAL', payload: true }
    ]

    expect(actions).toEqual(expectedPayload);
  });


  test('it should execute dispatch properly for event reducer', () => {

    const selectEvent = wrapperMount.find(`[data-testid="select-event-${day.format('DD-MM-YY')}"]`);
    selectEvent.simulate('click');


    const actions = store.getActions();
    const expectedPayload = [
      { "payload": dayjs("2021-11-10T08:10:00.000Z"), "type": "DAY/SET_DAY_SELECTED" },
      { "payload": true, "type": "MODAL/SET_MODAL" },
      {
        "payload": {
          "_id": "619f90f7abfdf7635dc91f43",
          "day": 1636531800000,
          "description": "Meet with friends",
          "id": "619f90f7abfdf7635dc91f43",
          "label": "purple",
          "title": "Meeting",
        }, "type": "EVENT/SET_SELECTED_EVENT",
      },
      { "payload": dayjs("2021-11-10T08:10:00.000Z"), "type": "DAY/SET_DAY_SELECTED" },
      { "payload": true, "type": "MODAL/SET_MODAL" },

    ]

    expect(actions).toEqual(expectedPayload);
  });




});
