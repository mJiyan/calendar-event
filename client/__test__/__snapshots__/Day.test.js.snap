// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering components it should match with the month component 1`] = `
<div
  className="border border-gray-200 flex flex-col"
>
  <header
    className="flex flex-col items-center"
    data-testid="day-header-10-11-21"
  >
    <p
      className="text-sm p-1 my-1 text-center "
      data-testid="day-value-10-11-21"
    >
      10
    </p>
  </header>
  <div
    className="flex-1 cursor-pointer"
    data-testid="select-day-10-11-21"
    onClick={[Function]}
  />
</div>
`;
